contributors/design-proposals/cri-dockershim-checkpoint.md

```diff
@@ -0,0 +1,123 @@
+the same information from PodSandboxConfig at creation and deletion. 
+
+```
+Kubelet ---------------------------------
+     | RunPodSandbox(PodSandboxConfig)
+     | StopPodSandbox(PodSandboxID)
+     V
+Dockershim-------------------------------
+     | SetUpPod
+     | TearDownPod
+     V
+Network Plugin---------------------------
+     | ADD
+     | DEL
+     V
+CNI plugin-------------------------------
+```
+
+
+In addition, checkpoint helps to improve the reliability of dockershim. 
+With checkpoints, critical information for disaster recovery could be 
+preserved. Kubelet makes decisions based on the reported pod states 
+from runtime shims. Dockershim currently gathers states from docker 
+engine. However, in the case of disaster recovery, docker engine may 
+lose all container information, including the reference ids. Without 
+the reference id, kubelet or dockershim could not reclaim allocated 
+resource. For example, if docker containers are removed underneath 
+kubelet, reference to the allocated IPs and iptables setup for the 
+pods are also lost. Hence leading to resource leak.  
+
+### Why checkpoint in dockershim?
+- CNI specification does not require CNI plugins to be stateful. And CNI 
+specification does not provide interface to retrieve states from CNI plugins. 
+- Currently there is no uniform checkpoint requirements across existing runtime shims.
+- Need to preserve backward compatibility for kubelet.
+- Easier to maintain backward compatibility by checkpointing at a lower level.
+
+## PodSandbox Checkpoint 
+Checkpoint file will be created for each PodSandbox. Files will be 
+placed under `/var/lib/dockershim/sandbox/`. File name will be `PodSandboxID`. 
+File content will be json encoded. Data structure is as follows:
+
+```go
+const schemaVersion = "v1"
+const PortMappingsKey = "PortMappings"
+
+type Protocol string
+
+// PortMapping is the port mapping configurations of a sandbox.
+type PortMapping struct {
+	// Protocol of the port mapping.
+	Protocol *Protocol `json:"protocol,omitempty"`
+	// Port number within the container.
+	ContainerPort *int32 `json:"container_port,omitempty"`
+	// Port number on the host.
+	HostPort *int32 `json:"host_port,omitempty"`
+}
+
+// PodSandboxCheckpoint is the checkpoint structure for a sandbox
+type PodSandboxCheckpoint struct{
+	// Version of the pod sandbox checkpoint schema.
+	Version string `json:"version"`
+	// Pod name of the sandbox. Same as the pod name in the PodSpec.
+	Name string `json:"name"`
+	// Pod namespace of the sandbox. Same as the pod namespace in the PodSpec.
+	Namespace string `json:"namespace"`
+	// Data to checkpoint for pod sandbox.
+	Data map[string]interface{} `json:"data,omitempty"`
+}
+
+```
+
+
+## Workflow Changes
+
+
+`RunPodSandbox` creates checkpoint: 
+```
+() --> Pull Image --> Create Sandbox Container --> (Create Sandbox Checkpoint) --> Start Sandbox Container --> Set Up Network --> ()
```

Comment:
Yes. `Create Sandbox Container` means `docker create ....`. After this, the sandbox id is returned, but sandbox container is not running. Cgroups and namespaces are not present at this point.